code/game/objects/structures/shelves.dm

```diff
@@ -0,0 +1,139 @@
+GLOBAL_LIST_INIT(shelf_colors, list("basic", "sci", "sup", "serv", "med", "sec", "cmd", "engi"))
+
+/obj/structure/shelf
+	name = "shelf"
+	desc = "A piece of furniture used for vertical storage."
+	icon = 'icons/obj/structures/shelves.dmi'
+	icon_state = "shelf_basic"
+	layer = TABLE_LAYER
+	density = TRUE
+	anchored = TRUE
+	pass_flags_self = PASSTAKE
+	max_integrity = 80
+	var/shelf_style = "basic"
+	var/image/shelf_overlay
+	COOLDOWN_DECLARE(spraypaint_cd)
+
+/obj/structure/shelf/Initialize(mapload)
+	. = ..()
+	var/area/A = get_area(src)
+	AddComponent(/datum/component/shelver/basic_shelf, random_pickup_locations_ = istype(A, /area/station/maintenance))
+	update_icon()
+	set_style(shelf_style)
+
+	if(mapload)
+		SEND_SIGNAL(src, COMSIG_SHELF_ADDED_ON_MAPLOAD)
+
+/obj/structure/shelf/attackby(obj/item/I, mob/living/user, params)
+	var/obj/item/toy/crayon/spraycan/spraycan = I
+	if(!istype(spraycan))
+		return ..()
+	if(spraycan.capped)
+		return ..()
+
+	if(!COOLDOWN_FINISHED(src, spraypaint_cd))
+		to_chat(user, "<span class='warning'>The paint on [src] is still drying!</span>")
+		return
+
+	var/cur_idx = GLOB.shelf_colors.Find(shelf_style)
+	if(!cur_idx)
+		cur_idx = 1
+	cur_idx++
+	if(cur_idx > length(GLOB.shelf_colors))
+		cur_idx = 1
+	set_style(GLOB.shelf_colors[cur_idx])
+	spraycan.play_spray_sound(user)
+
+	// Tiny cooldown to prevent constant spamming of the action
+	COOLDOWN_START(src, spraypaint_cd, 1 SECONDS)
+
+/obj/structure/shelf/proc/set_style(new_style)
+	if(shelf_style == new_style && shelf_overlay != null)
+		return
+	shelf_style = new_style
+	icon_state = "shelf_[shelf_style]"
+	// Don't think I can get away with both icon smoothing across tiles
+	// and dealing with this overlay shit
+	shelf_overlay = mutable_appearance('icons/obj/structures/shelves.dmi', "shelf_[shelf_style]-top")
+	shelf_overlay.layer = LOW_ITEM_LAYER
+	update_icon()
+
+/obj/structure/shelf/update_icon()
+	cut_overlays()
+	add_overlay(shelf_overlay)
```

Comment:
I think there might be `update_overlays()`? I'm not entirely sure, but if that's the case, you should probably put this logic under that and, when you want to update overlays, call update_appearance(UPDATE_OVERLAYS) or something.

Reply:
This seems to work as expected, PTAL